src/execution_plan/execution_plan.c

```diff
@@ -875,42 +860,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
 	array_free(segment_indices);
 
 	ExecutionPlan *plan = segments[segment_count - 1];
-
 	// The root operation is OpResults only if the query culminates in a RETURN or CALL clause.
-	if(query_has_return) {
-		if(!connecting_op) {
-			// Set the connecting op if our query is just a RETURN.
-			assert(segment_count == 1);
-			connecting_op = ExecutionPlan_LocateFirstOp(plan->root, OPType_PROJECT | OPType_AGGREGATE);
-		}
+	if(query_has_return || last_clause_type == CYPHER_AST_CALL) {
+		OpBase *results_op = NewResultsOp(plan);
+		_ExecutionPlan_UpdateRoot(plan, results_op);
+	}
 
-		// Prepare column names for the ResultSet.
-		if(result_set) ResultSet_SetColumns(result_set, AST_BuildColumnNames(last_clause));
+	// Disregard self.
+	plan->segment_count = segment_count - 1;
+	plan->segments = segments;
 
-		OpBase *results_op = NewResultsOp(plan, result_set);
-		_ExecutionPlan_UpdateRoot(plan, results_op);
+	return plan;
+}
+
+static void _ExecutionPlan_SetResultColumns(ExecutionPlan *plan) {
+	ResultSet *result_set = QueryCtx_GetResultSet();
+	AST *ast = QueryCtx_GetAST();
+	uint clause_count = cypher_ast_query_nclauses(ast->root);
+	const cypher_astnode_t *last_clause = cypher_ast_query_get_clause(ast->root, clause_count - 1);
+	cypher_astnode_type_t last_clause_type = cypher_astnode_type(last_clause);
+	bool query_has_return = (last_clause_type == CYPHER_AST_RETURN);
+	if(query_has_return) {
+		ResultSet_SetColumns(AST_BuildColumnNames(last_clause));
 	} else if(last_clause_type == CYPHER_AST_CALL) {
-		assert(plan->root->type == OPType_PROC_CALL);
-		OpProcCall *last_op = (OpProcCall *)plan->root;
+		OpBase *proc_call_base = ExecutionPlan_LocateFirstOp(plan->root, OPType_PROC_CALL);
+		assert(proc_call_base && (proc_call_base->type == OPType_PROC_CALL));
```

Comment:
`proc_call_base->type == OPType_PROC_CALL` redundant, you've asked for: `ExecutionPlan_LocateFirstOp(plan->root, OPType_PROC_CALL);`